src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -9088,6 +9088,20 @@ bool MergeTreeData::initializeDiskOnConfigChange(const std::set<String> & new_ad
     return true;
 }
 
+void MergeTreeData::loadPrimaryKeys()
+{
+    static DataPartStates affordable_states = { MergeTreeDataPartState::Active, MergeTreeDataPartState::Outdated, MergeTreeDataPartState::Deleting };
```

Comment:
We shouldn't load indexes for `Outdated` parts and especially for `Deleting`, because it may cause a race with removing of index file: https://s3.amazonaws.com/clickhouse-test-reports/0/d8a235b78557c3f731352b83b490f13e50fa90de/stateless_tests__azure__asan__[2_3].html

Reply:
https://github.com/ClickHouse/ClickHouse/pull/72614

Reply:
It is worse. When I ran the test locally, it sometimes got wrong results: the test expects the PK to be loaded but it isn't. I don't really had a quick explanation for that. Turned out that the same happens in master.

--> https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgMQogICAgLS0gQU5EIHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwCiAgICBBTkQgdGVzdF9zdGF0dXMgSU4gKCdGQUlMJywgJ0VSUk9SJywgJ0JST0tFTicpCiAgICBBTkQgdG9EYXRlKGNoZWNrX3N0YXJ0X3RpbWUpID4gJzIwMjQtMTEtMDEnCiAgICBBTkQgdGVzdF9uYW1lIGxpa2UgJyUwMzIwMl9zeXN0ZW1fbG9hZF9wcmltYXJ5X2tleSUnCk9SREVSIEJZIGNoZWNrX3N0YXJ0X3RpbWUgREVTQw==

Let's revert.

Reply:
It was a race with enabling randomization of setting `use_primary_key_cache` in master. #72615.

Reply:
Phew! :disappointed_relieved:

Reply:
https://github.com/ClickHouse/ClickHouse/pull/72628 to disable random settings